### ISSUE TITLE
Proposal for chained property syntax

### ISSUE THREAD

**Issue by chrisprice** [ScottLogic/d3-financial-components#275]
In #232 we introduced a new convention for the property accessors. Having spent some time with it, I really don't believe this new approach enhances readability (any clarity in defining the property is offset by the verboseness of every access). Therefore I propose we adopt one of the following syntaxes -

# d3

## Plain
```js
d3.layout.histogram = function() {
  var frequency = true,
      valuer = Number,
      ranger = d3_layout_histogramRange,
      binner = d3_layout_histogramBinSturges;
  
  // ...

  histogram.value = function(x) {
    if (!arguments.length) return valuer;
    valuer = x;
    return histogram;
  };
```

## Functor
```js
d3.layout.histogram = function() {
  var frequency = true,
      valuer = Number,
      ranger = d3_layout_histogramRange,
      binner = d3_layout_histogramBinSturges;
  
  // ...

  histogram.range = function(x) {
    if (!arguments.length) return ranger;
    ranger = d3_functor(x);
    return histogram;
  };
```

# Alternative terse syntax 

## Plain
```js
d3.layout.histogram = function() {
  var frequency = true,
      valuer = Number,
      ranger = d3_layout_histogramRange,
      binner = d3_layout_histogramBinSturges;
  
  // ...

  histogram.value = function(x) {
    return arguments.length ? (valuer = x, histogram) : valuer;
  };
```

## Functor
```js
d3.layout.histogram = function() {
  var frequency = true,
      valuer = Number,
      ranger = d3_layout_histogramRange,
      binner = d3_layout_histogramBinSturges;
  
  // ...

  histogram.range = function(x) {
    return arguments.length ? (ranger = d3_functor(x), histogram) : ranger;
  };
```

**Comment by chrisprice** [comment]
I forgot to mention the big win of these approaches, side effects can now easily be incorporated into the setter rather than being applied as part of the called function e.g. -

## Old
```js
       var bollingerBands = function(selection) {

            algorithm.inputValue(bollingerBands.yValue.value)
                .outputValue(bollingerBands.writeCalculatedValue.value);

            area.xScale(bollingerBands.xScale.value)
                .yScale(bollingerBands.yScale.value)
                .xValue(bollingerBands.xValue.value);

            upperLine.xScale(bollingerBands.xScale.value)
                .yScale(bollingerBands.yScale.value)
                .xValue(bollingerBands.xValue.value);

            averageLine.xScale(bollingerBands.xScale.value)
                .yScale(bollingerBands.yScale.value)
                .xValue(bollingerBands.xValue.value);

            lowerLine.xScale(bollingerBands.xScale.value)
                .yScale(bollingerBands.yScale.value)
                .xValue(bollingerBands.xValue.value);
            // ...

    bollingerBands.xScale = fc.utilities.property(d3.time.scale());
```
## New
```js
    bollingerBands.xScale = function(x) {
          if (!arguments.length) {
              return xScale;
          }
          xScale = x;
          area.xScale(x);
          upperLine.xScale(x);
          averageLine.xScale(x);
          lowerLine.xScale(x);
          return bollingerBands;
    };
```

**Comment by jleft** [comment]
:+1:  Agree with all of the above :)

I don't have a strong preference between the syntax options; possibly a slight preference for the _d3_ options, as I find it a bit easier to read, but there's not much in it!

**Comment by ColinEberhardt** [comment]
Agree with the general concept, I have a slight preference for the more terse syntax.

**Comment by ColinEberhardt** [comment]
What are your thoughts on macros? Basically use something like sweet.js to define the pattern just once. We always build our js code, so this should be viable.

**Comment by chrisprice** [comment]
If we're considering babel to be off the cards, then I think sweet.js would have to be as well. Having said that I'd happily see both in the project, along with a proper imports.

**Comment by chrisprice** [comment]
We've discussed this offline and in order to allow the expected usage pattern, we've agreed not to do anything to the code which makes it awkward to copy/paste components. That means no es6 modules (requires transpiler and means relative paths will break) and no macros (requires macro pre-processor).

Therefore I've implemented (#342) the properties like so -
```js
        measure.xLabel = function(x) {
            if (!arguments.length) {
                return xLabel;
            }
            xLabel = d3.functor(x);
            return measure;
        };
```

This is the closest to d3's implementation I could get whilst conforming to our style rules.
